etna/lib/etna/auth.rb

```diff
@@ -141,10 +201,19 @@ def approve_user(request)
       begin
         payload, header = application.sign.jwt_decode(token)
 
+        payload = symbolize_payload_keys(payload)
+
         return false unless janus_approved?(payload, token, request)
+
+        # If the user has not signed the code of conduct for a community
+        #   project, redirect them.
+        cc_required?(payload, token, request)
```

Comment:
I'm going to be honest -- I don't like how this logic works.  We're introducing an exception whose job is merely to overload the return value (true, false, do other unrelated thing), but also complicates control flow (is user authorized?  hold on, do some other thing by excepting up to a rescue block way above).

Part of it is that I think our etna auth library is getting bulky and confusing under its own weight and it's worth taking a look at some small refactors / break down to peel parts away more concertedly.

first off, I think that the cc redirect logic should entirely fit within fail or redirect.  Basically, in the original redirect, it should simply be able to determine if there is the possibility of a cc signage helping the user to login.  That should be the nature result of not being authorized by some other means.

Reply:
ah, taking that back, I think I see why you had to take this approach.

`approve_user` isn't actually responsible for authorizing, ONLY authenticating.

Maybe this logic ought to exist somewhere else (authorization layer, not authentication layer)

Reply:
Ok -- here's my ask.  See if you can move the cc redirect logic into `route.rb` around the `user_authorized` call.

You can not only check if authorization is failing before executing the redirect, you can capture the project_name sensibly and issue the redirect based on that.